components/script/dom/bindings/utils.rs

```diff
@@ -621,3 +633,93 @@ impl AsCCharPtrPtr for [u8] {
         self as *const [u8] as *const c_char
     }
 }
+
+pub unsafe fn html_constructor_handling<T: DerivedFrom<Element> + DomObject>(
+    cx: SafeJSContext,
+    args: CallArgs,
+    global: &Window,
+    get_proto_object: fn(SafeJSContext, HandleObject, MutableHandleObject),
+) -> bool {
+    // Step 2 https://html.spec.whatwg.org/multipage/#htmlconstructor
+    // The custom element definition cannot use an element interface as its constructor
+
+    // The new_target might be a cross-compartment wrapper. Get the underlying object
+    // so we can do the spec's object-identity checks.
+    rooted!(in(*cx) let new_target = UnwrapObjectDynamic(args.new_target().to_object(), *cx, 1));
+    if new_target.is_null() {
+        throw_dom_exception(
+            cx,
+            global.upcast::<GlobalScope>(),
+            Error::Type("new.target is null".to_owned()),
+        );
+        return false;
+    }
+
+    if args.callee() == new_target.get() {
+        throw_dom_exception(
+            cx,
+            global.upcast::<GlobalScope>(),
+            Error::Type("new.target must not be the active function object".to_owned()),
+        );
+        return false;
+    }
+
+    // Step 6
+    rooted!(in(*cx) let mut prototype = ptr::null_mut::<JSObject>());
+    {
+        rooted!(in(*cx) let mut proto_val = UndefinedValue());
+        let _ac = JSAutoRealm::new(*cx, new_target.get());
+        if !JS_GetProperty(
+            *cx,
+            new_target.handle(),
+            b"prototype\\0".as_ptr() as *const _,
+            proto_val.handle_mut(),
+        ) {
+            return false;
+        }
+
+        if !proto_val.is_object() {
+            // Step 7 of https://html.spec.whatwg.org/multipage/#htmlconstructor.
+            // This fallback behavior is designed to match analogous behavior for the
+            // JavaScript built-ins. So we enter the compartment of our underlying
+            // newTarget object and fall back to the prototype object from that global.
+            // XXX The spec says to use GetFunctionRealm(), which is not actually
+            // the same thing as what we have here (e.g. in the case of scripted callable proxies
+            // whose target is not same-compartment with the proxy, or bound functions, etc).
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1317658
+            rooted!(in(*cx) let global_object = CurrentGlobalOrNull(*cx));
+            get_proto_object(cx, global_object.handle(), prototype.handle_mut());
+        } else {
+            // Step 6
+            prototype.set(proto_val.to_object());
+        };
+    }
+
+    // Wrap prototype in this context since it is from the newTarget compartment
+    if !JS_WrapObject(*cx, RawMutableHandle::from(prototype.handle_mut())) {
+        return false;
+    }
+
+    let result: Result<DomRoot<T>, Error> = html_constructor(&global, &args);
```

Comment:
Since this is the only code that is generic in this method, let's split the code up like this:
```rust
fn prepare_html_constructor_prototype(cx: SafeJSContext, args: CallArgs, global: &Window, get_proto_object: fn(SafeJSContext, HandleObject, MutableHandleObject), prototype: MutableHandleObject) -> Result<(), Error> {
    // more or less everything before the html_constructor call
}

fn setup_html_constructor_prototype(cx: SafeJSContext, reflector: HandleObject, prototype: HandleObject, wrapped: MutableHandleValue) {
    // everything after the html_constructor call
}

fn call_html_constructor<T: ...>(...) -> bool {
    rooted!(in(*cx) let mut prototype = ptr::null_mut::<JSObject>());
    if let Err(e) = prepare_html_constructor_prototype(..., prototype.handle_mut()) {
        throw_dom_exception(cx, global.upcast(), e);
        return false;
    }
    // call html_constructor, store in result
    setup_html_constructor_prototype(cx, result.reflector().get_jsobject(), MutableHandleValue::from_raw(args.rval())).is_ok()
}
```
This should avoid generating lots of code that gets monomorphized unnecessarily during the build.

Reply:
In principle I understand that monomorphizing anything that is not needed adds complexity and slows down compile times. Would you say these are the main reasons for avoiding monomorphizing code unnecessarily?